pep.rst

```diff
@@ -0,0 +1,295 @@
+          }
+      }
+
+     if (code->co_run_count++ > PyJIT_HOT_CODE) {
+         PyThreadState *tstate = PyThreadState_GET();
+         PyInterpreterState *interp = tstate->interp;
+         if (interp->jit_compile != NULL) {
+             code->co_jitted = interp->jit_compile((PyObject*)code);
+             if (code->co_jitted != NULL && code->co_jitted != PY_JIT_FAILED) {
+                 // Compilation succeeded!
+                 return code->co_jitted->jit_evalfunc(code->co_jitted->jit_evalstate, f);
+             }
+
+             // Compilation failed, so no longer try and compile this
+             // method.
+             code->co_compilefailed = PY_JIT_FAILED;
+         }
+     }
+
+     // Fall-through; use CPython's normal eval loop.
+     return PyEval_EvalFrameEx_NoJit(f, throwflag);
+
+
+Implementation
+==============
+
+A set of patches implementing the proposed API is available through
+the Pyjion project [#pyjion]_. The project also includes a
+proof-of-concept JIT using the CoreCLR JIT [#coreclr]_ (called
+RyuJIT).
+
+
+Open Issues
+===========
+
+Provisionally accept the proposed changes
+-----------------------------------------
+
+While PEP 411 introduced the concept of provisionally accepted
+packages in Python's standard library, the concept has yet to be
+applied to CPython's C API. Due to the unknown payoff from adding this
+API to CPython, it may make sense to provisionally accept this PEP
+with a goal to validate its usefulness based on whether JITs emerge
+which make use of the proposed API.
+
+
+Make the proposed API a compile-time option
+-------------------------------------------
+
+While the API is small and performance impact of executions with no
+JIT in use should be minimal (the default, no-JIT case consists of
+1 ``!=`` comparison, a ``>`` comparison, and an increment), there will
+always be some overhead. If the C API is deemed worth having but the
+performance cost in the non-JIT case is considered too high, the API
+could become a compile-time option. This is obviously not preferred as
+it makes it more of a burden to use the new C API.
+
+
+How to specify what JIT to use?
+-------------------------------
+
+Should JITs be an explicit ``-X`` flag for CPython? Or should a JIT
+simply be like any other extension module that gets imported and it is
+up to the module to register the necessary functions during module
+initialization?
+
+
+
+Rejected Ideas
+==============
+
+A separate boolean to flag when a code object cannot be compiled
+----------------------------------------------------------------
+
+In the first proof-of-concept of the proposed API there was a
+``co_compilefailed`` flag on code objects that was set by the JIT when
+it was unable to compile the code object. This was eventually removed
+as it was deemed unnecessary when ``co_jitted`` could be assigned a
+constant value for the same purpose, eliminating the need for memory
+per code object just for this flag.
+
+
```

Comment:
An earlier version only added an evaluation function to the code object and did not expose a JittedCode object.  This had a couple of limitations.  First, there was no way for a JIT to put in place a tracing function which could be shared across multiple code objects.  Second, the expose of the JittedCode object will allow JITs to create decorators which could provide hints to the JIT to control code generation.  For example a JIT could have a decorator to force or prevent inlining.